internal/docker/client.go

```diff
@@ -645,3 +649,99 @@ func (c *Cli) ExecInContainer(ctx context.Context, cID container.ID, cmd model.C
 		return nil
 	}
 }
+
+func (c *Cli) Run(ctx context.Context, opts RunOptions) (RunResult, error) {
+	image := opts.ContainerConfig.Image
+	if image == "" {
+		// docker doesn't give great errors, so we do some basic validation upfront
+		return RunResult{}, errors.New("image name (opts.ContainerConfig.Image) cannot be empty")
+	}
+
+	if opts.Pull {
+		ref, err := reference.ParseNamed(image)
+		if err != nil {
+			return RunResult{}, fmt.Errorf("invalid image ref %q: %v", image, err)
+		}
+		if _, err := c.ImagePull(ctx, ref); err != nil {
+			return RunResult{}, fmt.Errorf("error pulling image %q: %v", image, err)
+		}
+	}
+
+	createResp, err := c.Client.ContainerCreate(ctx,
+		&opts.ContainerConfig,
+		opts.HostConfig,
+		opts.NetworkConfig,
+		nil,
+		opts.ContainerName,
+	)
+	if err != nil {
+		return RunResult{}, fmt.Errorf("could not create container: %v", err)
+	}
+	var containerStarted bool
+	defer func() {
+		// make an effort to clean up any container we create but don't successfully start
+		if containerStarted {
+			return
+		}
+		if err := c.Client.ContainerRemove(ctx, createResp.ID, types.ContainerRemoveOptions{}); err != nil {
```

Comment:
do we also need to remove the container when it exits? maybe this should be a Close() method of the RunResult?

Reply:
I like that - combined with reducing surface area of options it makes it harder to misuse. Added in f758868